### ISSUE TITLE
rvc as an interface to edit connection configuration

### ISSUE THREAD

**Issue by ronaldtse** [riboseinc/rvc#94]
(Open to suggestions!)

## Connection options

```sh
rvc connection set [vpn-name] autoconnect [value]
rvc connection set [vpn-name] name [new-vpn-name]
rvc connection set [vpn-name] pre-connect-exec [/script/to/exec]
rvc connection set [vpn-name] post-connect-exec [/script/to/exec]
rvc connection set [vpn-name] ca-cert [/path/to/ca.crt]
rvc connection set [vpn-name] client-cert [/path/to/client.crt]
rvc connection set [vpn-name] client-key [/path/to/client.key]
rvc connection set [vpn-name] ovpn-config [/path/to/client.ovpn]
```

## Import Connection

```sh
rvc connection new-from-tblk [vpn-name] [/path/to/client.tblk]
rvc connection new-from-ovpn [vpn-name] [/path/to/foobar.ovpn]
```

## Add Connection

```sh
rvc connection add [vpn-name]
```
```
WARNING: [vpn-name] already exists, re-run with `--force` to replace.
```

## Remove Connection

```sh
rvc connection remove [vpn-name] [/path/to/foobar.ovpn] (--force to execute)
```
```
WARNING: This is a dry-run. Re-run with `--force` to remove connection.
```

## View options

```sh
rvc connection show [vpn-name]
rvc connection show [vpn-name] --json
```

**Comment by erikbor** [comment]
My comments one by one each subject:

## Connection options

These options look good and modify **only** the corresponding `.json` file of `vpn-name`:
```console
$ sudo rvc connection set [vpn-name] autoconnect <value>
$ sudo rvc connection set [vpn-name] name <new-vpn-name>
$ sudo rvc connection set [vpn-name] pre-connect-exec </script/to/exec>
$ sudo rvc connection set [vpn-name] post-connect-exec </script/to/exec>
```

The following options would edit a `.ovpn` file. I think given all the options that OpenVPN offers it would be just a too limited set that we support and therefor I think we should not support editing `.ovpn` files at all. Plus it would be a nightmare to check for all the permissions of each of the files specified in the arguments. @ronaldtse @jjr840430 what do you think?
```
$ sudo rvc connection set [vpn-name] ca-cert [/path/to/ca.crt]
$ sudo rvc connection set [vpn-name] client-cert [/path/to/client.crt]
$ sudo rvc connection set [vpn-name] client-key [/path/to/client.key]
$ sudo rvc connection set [vpn-name] ovpn-config [/path/to/client.ovpn]
```

## Import Connection
See [95](https://github.com/riboseinc/rvc/issues/95)

## Add Connection
I think `add` is basically a duplicate of `import` which is [95](https://github.com/riboseinc/rvc/issues/95) plus [rvc reload 93](https://github.com/riboseinc/rvc/issues/93)

Or is there any difference that I don't see?

## Remove Connection
So remove connections by name and `.ovpn` files in `/opt/rvc/etc/vpn.d`. When removing, do we `unlink()`? I suppose we also need a reload of `rvd`.

**Comment by ronaldtse** [comment]
@erikbor there is actually one more use case I thought we should support: the incomplete `ovpn` file.

By "incomplete ovpn file" I mean the separation of the `.ovpn` file (OpenVPN config), the `ca.crt` (CA Cert), the `client.crt` (Client Cert) and `client.key` (Client Key). This is the case in Tunnelblick.

The "complete ovpn file" contains all 4 elements inside the `.ovpn` file.

I added the second set of "connection set" options because of "Add Connection".

For example, if I was given these 4 files separately, I could do this:
```sh
$ sudo rvc connection add [vpn-name]
$ sudo rvc connection set [vpn-name] ca-cert [/path/to/ca.crt]
$ sudo rvc connection set [vpn-name] client-cert [/path/to/client.crt]
$ sudo rvc connection set [vpn-name] client-key [/path/to/client.key]
$ sudo rvc connection set [vpn-name] ovpn-config [/path/to/client.ovpn]
```

I suppose if we allow this, we should store all of this in an internal state (e.g., JSON) and then only serialize out to an `.ovpn` file when we use OpenVPN to connect. The good thing is if we do this, we could potentially be independent of OpenVPN (in the future)?

**Comment by ronaldtse** [comment]
OK I spoke with @erikbor and the priority is to support only the first set of configs, since we won't be switching out of OpenVPN in the near future.

**Comment by jjr840430** [comment]
@erikbor `rvc connect set` and `rvc connection new*` will require sudo.
Even if rvc executes it by sudo privilege, but rvd couldn't check whether that command is from sudo process.
So if someone knows about the sending format of above commands, then it's possible to apply them using the normal user by sending the same format to rvd process.
What're your thoughts on that?

**Comment by jjr840430** [comment]
@ronaldtse 
```
$ sudo rvc connection add [vpn-name]
$ sudo rvc connection set [vpn-name] ca-cert [/path/to/ca.crt]
$ sudo rvc connection set [vpn-name] client-cert [/path/to/client.crt]
$ sudo rvc connection set [vpn-name] client-key [/path/to/client.key]
$ sudo rvc connection set [vpn-name] ovpn-config [/path/to/client.ovpn]
```
It's possible to overwrite the contents of certificates and key files in OpenVPN configuration file.
So if `ovpn-config` option was set at first, then certs and key may be added in the ovpn config file.

**Comment by ronaldtse** [comment]
@jjr840430 the `rvc connection new*` and `rvc connection set` commands directly edit the connection's config JSON without going through `rvd`, so the permissions work just like before. 

It is only after changing the config it will ask `rvd` through the socket to re-check the config. Thanks!

**Comment by jjr840430** [comment]
Implemented remove_connection by #115

**Comment by jjr840430** [comment]
@ronaldtse @erikbor Here is the command to edit `rvc` connections.

```
sudo rvc  edit <connection name> <connection params> <parameter values>
```

Here is the list of `connection parameters`:

`auto-connect` : The `flag` for auto connection. The value may be `true` or `false`.
`pre-exec-cmd` : The `command line` which is triggered before connecting to `VPN`.
`profile` : The `profile` path for `VPN connection`.
`certificate` : `PKCS#12` certificate including `CA`, `client certificate` and `private key`.
`keychain-item` : `keychain` item contains `client certificate` and `private key`

Please let me know your opinions.

**Comment by ronaldtse** [comment]
@jjr840430 do you mean that `keychain-item` would allow us to store those parts in the macOS keychain? Is it possible to "import" an `ovpn` file and automatically have those parts stored in the keychain?

**Comment by jjr840430** [comment]
If `keychain-item` option is used, then the credentials in original `ovpn` file will be replaced to the credentials which are stored in `keychain`
